Microsoft.AVS.Management/Microsoft.AVS.Management.psm1

```diff
@@ -880,27 +907,20 @@ function Set-AvsVMStoragePolicy {
     } 
 
     $ProtectedVMs = Get-ProtectedVMs 
-    if ($ProtectedVMs.Name.Contains($VMName)) {
-        Write-Error "Access denied to this VM." -ErrorAction Stop
-    }
-    $VM = Get-VM $VMName
-    if ($null -eq $VM) {
-        Write-Error "Was not able to set the storage policy on the VM. Could not find VM with the name: $VMName" -ErrorAction Stop
-    }
-    if (-not $(Get-SpbmEntityConfiguration $VM).StoragePolicy -in $VSANStoragePolicies) {
-        Write-Error "Modifying Storage policy on this VM is not supported" -ErrorAction Stop
-    }
-    Write-Host "Setting VM $VMName storage policy to $StoragePolicyName..."
-    try {
-        Set-VM -VM $VM -StoragePolicy $StoragePolicy -ErrorAction Stop -Confirm:$false
-    }
-    catch [VMware.VimAutomation.ViCore.Types.V1.ErrorHandling.InvalidVmConfig] {
-        Write-Error "The selected storage policy $($StoragePolicy.Name) is not compatible with this VM. You may need more hosts: $($PSItem.Exception.Message)" -ErrorAction Stop
-    }
-    catch {
-        Write-Error "Was not able to set the storage policy on the VM: $($PSItem.Exception.Message)" -ErrorAction Stop
-    }
-    Write-Output "Successfully set the storage policy on VM $VMName to $StoragePolicyName"
+    $VMList = Get-VM $VMName | Where-Object {-not $_.Name -in $ProtectedVMs.Name}
+
+    if ($null -eq $VMList) {
+        Write-Error "Was not able to set the storage policy on the VM. Could not find VM(s) with the name: $VMName" -ErrorAction Stop
+    } elseif ($VMList.count -eq 1) {
+        $ErrorActionPreference = "Stop"
+        $VM = $VMList[0]
+        Set-StoragePolicyOnVM -VM $VM -VSANStoragePolicies $VSANStoragePolicies -StoragePolicy $StoragePolicy -ErrorAction Continue
+    } else {
+        $ErrorActionPreference = "Continue"
+        foreach ($VM in $VMList) {
+            Set-StoragePolicyOnVM -VM $VM -VSANStoragePolicies $VSANStoragePolicies -StoragePolicy $StoragePolicy -ErrorAction Stop
+        }
```

Comment:
$ErrorActionPreference defaults to 'Continue', so I think you could refactor here to only set $ErrorActionPreference in the case of a single element array and eliminate the redundant Set-StoragePolicyOnVM calls and handle either case in the foreach by dropping the explicit -ErrorAction parameter as in examples here: https://docs.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_preference_variables?view=powershell-7.1#examples-2

Reply:
So the agent sets ErrorAction preference to Stop. So in the context of this execution, I think the default preference would be "Stop" since the agent calls essentially Set-AvsVMStoragePolicy -ErrorAction Stop

Reply:
I see what you mean. So then couldn't you set $ErrorActionPreference to Continue (as you already do) for size > 1 and run Set-AvsVMStoragePolicy w/o the -ErrorAction parameter inside the foreach for the case of any array size > 0? This looks like we're doing more work than we need to, e.g. explicitly grabbing first element of array if size is 1.